### ISSUE TITLE
PyPI stale caches

### ISSUE THREAD

**Issue by Xcelled** [pypa/bandersnatch#56]
At work, we're using a slight modified older version of Bandersnatch to mirror PyPI. About 10 days ago the mirroring job started failing - a lot - because the PyPI cache we were hitting was stale and refusing to update. After 10 days of failure, (where we got behind by about 4,000 updates), I finally forced Bandersnatch to hit a different cache by changing some request headers.

All was well for about 24 hours, then *that* cache started having stale issues. I've since updated our Bandersnatch code to use a query parameter of the current timestamp in order to unconditionally bust the caches.

That's obviously a sub-optimal solution, but it got us back up and running. It also got me thinking. instead of just bailing after a certain number of stale page errors, what if Bandersnatch fell back to using the cache-bust hack? It could retry 3 times to see if the caches get updated, and if they don't, it would force the cache to be busted in order to retrieve the proper version.

I'm willing to make a PR for this (I'm trying to upgrade our Bandersnatch at work), but I didn't want to do so without a discussion first. Maybe there's a better way to do this?

**Comment by cooperlees** [comment]
Hi,

Thanks for reporting this, I have also seen this before. Let me get some other PyPA Warehouse people to read this before we rush to implement your suggestion, I also want to think about this a little more.

Cooper

**Comment by Xcelled** [comment]
I can get the cache logs for you in a bit. The other modifications we made were just to upload the syncd packages to Amazon AWS. Can't imagine that's terribly useful to others.

**Comment by di** [comment]
Thanks for the report! The XML-RPC API is a special snowflake and handles caching a bit differently than everything else on PyPI, so it's possible this is occasionally failing and we are unaware.

I'm a bit surprised that adding a cache breaker URL param helps here, because it shouldn't have any effect on the cached response AFAIK.

@cooperlees I'm not too familiar with Bandersnatch, can we be specific about which method would be failing here? Is it `updated_releases`?

**Comment by Xcelled** [comment]
I can give a better report when I get back to my desk, but it's the call to fetch the package json that was failing. I believe the x-pypi-last-serial header was being cached improperly, as the fetched json had the new serial. Unfortunately bandersnatch checks the header.

**Comment by cooperlees** [comment]
Thanks for the excellent problem description and root causing. The people I tagged above can talk about the caching better, and hopefully give us a better solution.

@di - Bandersnatch only uses the following two XMLRPC calls:
- `list_pakages_with_serial()` 
- `changelog_since_last_serial(serial: int)`
Source: https://github.com/pypa/bandersnatch/blob/master/src/bandersnatch/master.py#L84

Ideally I’d love to get off XMLRPC as I stated @ PyCon. Can we resume the XMLRPC replacement work? I’m happy to help, but I lost movement with others.

Is there a cleaner workaround bandersnatch could do here or could we root cause the cache issues in the short term?

**Comment by di** [comment]
We occasionally see this (namely, the difference between a plain and gzip response), though not quite this bad or frequently. Unfortunately when this happens, it seems to be mostly out of our control and instead an issue with our CDN.

One thing Bandersnatch could do is purge the cache itself if it suspects the cache is stale, by issuing an `XPURGE` to the stale endpoint. This would clear both the plain and the gzipped cache.

**Comment by Xcelled** [comment]
@di can anyone purge the cache? If so I can add that to our codebase so we stop busting caches unneededly.

**Comment by di** [comment]
@Xcelled Yup, anyone can issue an XPURGE to any PyPI page or endpoint (except for XML-RPC).

**Comment by Xcelled** [comment]
I can make a PR for this as well.

**Comment by cooperlees** [comment]
Awesome. This sounds much cleaner. Once we merge this I’ll stop being lazy and release 3.0 to PyPI too. Thanks all!

**Comment by dstufft** [comment]
I would prefer it if clients did not purge PyPI itself. This should generally be treated as a bug in PyPI not something for bandersnatch to take care of.

**Comment by dstufft** [comment]
Eh, I mean technically yes? My main concern with working around it in Bandersnatch is that we otherwise likely won’t even know it’s broken if Bandersnatch is just “fixing” It for us by issuing the purge. We’ll just have degraded performance on requests that we won’t otherwise notice.

**Comment by cooperlees** [comment]
I feel ya @dstufft - I’ve opened  https://github.com/pypa/warehouse/issues/4892 so that we can follow up debugging and removing this.

Can we somehow get more logging about how often Bandersnatch is hitting this and then think about how from this we will try and debug why requests is hitting this ... Lets discuss on the warehouse issue and I’ll keep this open until we remove the #57 hack.
